### ISSUE TITLE
Is there a plan to implement streaming data processing?

### ISSUE THREAD

**Issue by dbalabka** [pola-rs/polars#3379]
Recently, I've faced a problem using a large dataset that can not be processed using Pandas DataFrame + Python because of a required excessive amount of RAM caused by created copies of the DataFrame. There are various solutions that can help to overcome this limitation like: manually created batch processing, applying more advanced tools like Apache Beam, and others. During my investigation, I've found the Polars library because it employs the Arrow format that supports data streaming and zero-copy. It was sound like a possible solution to my problem but as I realize later Polars does not replace batch/streaming processing tools.

The task was to consume some CSV/Parquet files from disk, process them in most memory and CPU efficient way to perform feature engineering, and save it back to disk in different or the same format in batches (50MB per file). The resulting files should be used as a dataset for model training (e.g., using Tensorflow).

Creating custom functions using Rust and zero-copy looks very attractive for data processing. I think that Arrow's and Rust's performance and memory usage efficiency can give a possibility to process very large datasets even on a single machine. Do you think that Polars can cover such use cases in the future?

**Comment by ritchie46** [comment]
Polars is a lot more memory efficient than pandas is, however polars is not streaming, nor will it ever be. The algorithms require all data in memory.

Thinks that may help though is using polars lazy. If you filter your data, these filter can be applied at the scan level and don't have to be materialized. This can reduce memory pressure a lot depending on selectivity.

**Comment by dbalabka** [comment]
@ritchie46 thanks for the quick reply. 

> however polars is not streaming, nor will it ever be. The algorithms require all data in memory.
I fully understand this limitation. 

> If you filter your data, these filter can be applied at the scan level and don't have to be materialized
Still, can be possible to implement a possibility to stream not yet materialized DataFrame directly into disk w/o keeping it in memory? It might solve the problem in my case.

**Comment by ritchie46** [comment]
> Still, can it be possible to implement a possibility to stream not yet materialized DataFrame directly into disk w/o creating a new DataFrame in memory? It might solve the problem in my case.

Where would that stream come from? Polars' algorithms do not output streams, but dataframes.

**Comment by dbalabka** [comment]
@ritchie46 makes sense. Thanks for your answer!

**Comment by ritchie46** [comment]
You're welcome :)

**Comment by dbalabka** [comment]
@ritchie46, hey. It has been a while since the time I submitted this ticket. From the recent [interview with Jetbrains](https://blog.jetbrains.com/dataspell/2023/08/polars-vs-pandas-what-s-the-difference/), I understood that streaming became one of the critical focuses for Polar Library. Also, I noticed that Streaming API became available:
https://pola-rs.github.io/polars-book/user-guide/concepts/streaming/

Is there any roadmap plan available online to follow the updates? :)
